### PR TITLE
refactor!(core): rework attribute semantics

### PR DESCRIPTION
**Pull request by imrn99**
Sorry for the size of the last 2 PRs! 

**New semantic over attributes**:
- try / regular / force for merges & splits (same as (un)sews, see #224)
- regular / force for RW accesses (same as (un)links, see #224)
- RW access methods are now limited to:
	- `read`: read a value
	- `write`: (over)write a value, returning the old one
	- `remove`: remove a value, returning it

**Done**:
- add a new enum, `CMapError`
- add type alias `CMapResult<T> = Result<T, CMapError>`
- update embed methods to use the new semantic
- update the `AttributeStorage<A>` & `UnknownAttributeStorage` to use the new semantic
	- also update the trait implementation for `AttrSparseVec`
- update `AttrStorageManager` to use the new semantic
- update the documentation with explanation over the new semantic

**Follow-up**:
- add tests over edge case allowed by the new semantics (e.g. a potential transaction cancel over some concurrent changes)
- replace some options with with `CMapResult<A>` where possible / coherent
	- this corresponds mostly to places with `CMapResult<Option<A>>`
- update `AttributeUpdate` fallback methods to return `CMapResult<Self>` 
	- this creates even more control, with some attributes being essential and others fail-proof

### CODE REVIEW

**Review by cedricchevalier19 — APPROVED** [review]
Looks ok, I will mainly trust the Rust toolchain on this PR.

File: honeycomb-core/src/cmap/dim2/embed.rs [hunk 1]
```diff
@@ -22,13 +24,18 @@ impl<T: CoordsFloat> CMap2<T> {
...
     ///
     /// - `vertex_id: VertexIdentifier` -- Identifier of the given vertex.
     ///
-    /// # Return
+    /// # Return / Errors
+    ///
+    /// This method is meant to be called in a context where the returned `Result` is used to
+    /// validate the transaction passed as argument. The result should not be processed manually,
+    /// only used via the `?` operator.
     ///
```
[the remark this answers is not in the document]
Comment (imrn99) [reply honeycomb-core/src/cmap/dim2/embed.rs:39]: I believe the `stm` crate uses these errors to detect invalid transaction (early?), so yes, the `?` operator should be called on every results inside a transaction, including `Result<(), StmError>`.

**Comment by imrn99** [comment]
> Looks ok, I will mainly trust the Rust toolchain on this PR.

We can set something up this week to go over the API and any questions / shortcomings.
